Capitolo sul ciclo del feedback (#212)
Il-Libro-Open-Source/book: docs/it/ciclo-del-feedback.md

Sicuramente è grazie al fatto che possiamo lavorare con strumenti frutto di decenni di evoluzione tecnologia esponenziale, ma la vera ragione, il vero discriminante, sta nel fatto nel "soft" in "software": il prodotto del nostro lavoro è spesso intangibile e facilmente manipolabile, tanto che se commettiamo un errore, non dobbiamo prendere il semilavorato e buttarlo via, sprecando materie prime, ma nella peggiore delle ipotesi abbiamo solo sprecato un po' di tempo.

Tutto questo non è nient'altro che la "Terza rivoluzione industriale", come viene chiamata nei libri di scuola; dopo alcuni decenni in cui il collo di bottiglia era la potenza di calcolo, al giorno d'oggi abbiamo a disposizione CPU e memoria sufficienti per qualunque task mondano, e questa pura forza bruta si schianta ora invece contro problemi organizzativi: far lavorare decine, se non centinaia di sviluppatori assieme sullo stesso progetto software è un problema non banale, e che tanti prima di noi si sono trovati a dover risolvere.

> Uno strumento che è emerso vincitore da queste difficoltà è il "ciclo del feedback"; ma di che si tratta? E' un approccio in cui, in seguito di una semplice azione, si va a misurarne rapidamente l'effetto (il "feedback"), così da ottenerne nuove informazioni, che saranno impiegate per intraprendere i prossimi passi in maniera più conscia e informata, creando così il "ciclo". Questo ciclo generalmente permette di procedere in una attività a piccoli passi ma con una discreta rapidità e, cosa ancora più importante, sempre nella giusta direzione, visto che eventuali errori di rotta possono essere rapidamente corretti ad ogni passo.

Reviewer: ```suggestion
Uno strumento che è emerso vincitore da queste difficoltà è il **ciclo del feedback**; ma di che si tratta? È un approccio in cui, a seguito di un'azione, si va a misurarne rapidamente l'effetto (il "feedback"), così da ottenerne nuove informazioni, che saranno impiegate per intraprendere i prossimi passi in maniera più conscia e informata, creando così un ciclo continuo. Questo generalmente permette di procedere in un'attività a piccoli passi ma con una discreta rapidità e, cosa ancora più importante, sempre nella giusta direzione, visto che eventuali errori di rotta possono essere rapidamente corretti ad ogni passo.
```

## Il manifesto Agile e l'Extreme Programming

> Alcuni movimenti che sono nati in seno alle difficoltà di questa rivoluzione sono il Manifesto Agile, l'Extreme Programming, il DevOps: il primo ad alto livello, il secondo ad un livello più "basso" tecnico, il terzo in maniera "mista", hanno tutti cercato di trovare modi per scrivere software in maniera più snella e organizzata, in modo che non ci si ritrovasse più in una melassa di linee di codice che, dopo pochi mesi di lavoro, diventava ingestibile.

Reviewer: ```suggestion
Alcuni movimenti che sono nati in seno alle difficoltà di questa rivoluzione sono il Manifesto Agile, l'Extreme Programming (XP), il DevOps: il primo ad alto livello (_descrittivo_), il secondo ad un livello più _basso_ (tecnico), il terzo di tipo "misto", hanno tutti cercato modi per scrivere software in maniera più snella e organizzata, in modo che non ci si ritrovasse più in una melassa di linee di codice che, dopo pochi mesi di lavoro, risultasse ingestibile.
```

Reviewer: L'ho provata a correggere qua e là ma, pur comprendendo il senso che vuoi dare al contenuto, la proverei a riscrivere. Mi sembra un periodo troppo lungo in cui ci sono tanti concetti che al lettore potrebbero non essere noti. Che ne dici di prendere più respiro e aprire ognuno di questi 3 movimenti con qualche accenno o nozione aggiuntiva?

Inoltre attenzione che se la sezione la chiamiamo "Il manifesto Agile e l'Extreme Programming" e poi parli anche di DevOps rischiamo di confondere il lettore.

Alle fondamenta di questi approcci si trova **il ciclo del feedback**: vista la malleabilità del software e la relativa semplicità con cui si può evolvere e sperimentare, avere un ciclo del feedback breve ed efficace si è rapidamente dimostrato lo strumento vincente per uscire da quell'impasse.

> ## Il TDD

Reviewer: Ridurrei lo scope di questa sezione, indicando al letture che nell'apposito capitolo **Testing** si troveranno più informazioni riguardo a questa metodologia

Author: Ok, ma pensi abbia senso tenere la descrizione del ciclo? Quello mi sembra un collegamento necessario per spiegare perché lo cito..

Reviewer: Absolutely, però non andrei in spiegazioni dettagliate perché, appunto, il capitolo sul testing ne parla ampiamente e ha senso rimandare il lettore lì piuttosto che raddoppiare un contenuto, però certo, senza dubbio è più che sensato metterlo!

---
layout: default
title: Ciclo del Feedback
nav_order: 999
---

# Il Ciclo del Feedback

## La flessibilità del software

Intuitivamente parlando, sembra ovvio che noi sviluppatori, con una connessione internet e un portatile alla mano, possiamo essere enormemente più produttivi di decine di operai da catena di montaggio del secolo scorso. Ma perché è davvero così?

> Sicuramente è grazie al fatto che possiamo lavorare con strumenti frutto di decenni di evoluzione tecnologia esponenziale, ma la vera ragione, il vero discriminante, sta nel fatto nel prefisso "soft" nella parola "software": il prodotto del nostro lavoro è spesso intangibile e facilmente manipolabile, tanto che se commettiamo un errore, non dobbiamo prendere il semilavorato e buttarlo via, sprecando materie prime, ma nella peggiore delle ipotesi abbiamo solo sprecato un po' di tempo.

Reviewer: ```suggestion
Sicuramente un fattore è la possibilità di lavorare con strumenti frutto di decenni di evoluzione tecnologica esponenziale, ma la vera ragione, il vero discriminante, sta nel prefisso "soft" nella parola "software": il prodotto del nostro lavoro è spesso intangibile e facilmente manipolabile, tanto che se commettiamo un errore, non dobbiamo prendere il semilavorato e buttarlo via, sprecando materie prime, ma nella peggiore delle ipotesi abbiamo solo sprecato un po' di tempo.
```

Tutto questo non è nient'altro che la "Terza rivoluzione industriale", come viene chiamata nei libri di scuola; dopo alcuni decenni in cui il collo di bottiglia è stata la potenza di calcolo, al giorno d'oggi abbiamo a disposizione CPU e memoria sufficienti per qualunque task mondano, ma questa pura forza bruta ha esposto una nuova tipologia di problemi, quelli organizzativi: far lavorare decine, se non centinaia di sviluppatori assieme sullo stesso progetto software è un problema non banale, e che tanti prima di noi si sono trovati a dover risolvere.

> Uno strumento che è emerso vincitore da queste difficoltà è il **ciclo del feedback**; ma di che si tratta? È un approccio in cui, a seguito di un'azione, si va a misurarne rapidamente l'effetto (il "feedback"), così da ottenerne nuove informazioni, che saranno impiegate per intraprendere i prossimi passi in maniera più conscia e informata, creando così un ciclo continuo. Questo generalmente permette di procedere in un'attività a piccoli passi ma con una discreta rapidità e, cosa ancora più importante, sempre nella giusta direzione, visto che eventuali errori di rotta possono essere rapidamente corretti ad ogni passo.

Reviewer: ```suggestion
Uno strumento che è emerso vincitore da queste difficoltà è il **ciclo del feedback**; ma di che si tratta? È un approccio in cui, a seguito di un'azione, si va a misurarne rapidamente l'effetto (il "feedback"), così da ottenerne nuove informazioni, che saranno impiegate per intraprendere i prossimi passi in maniera più conscia e informata, creando così un ciclo continuo. Questo generalmente permette di procedere in un'attività a piccoli passi ma con una discreta rapidità e, cosa ancora più importante, solitamente nella giusta direzione, visto che eventuali errori di rotta possono essere rapidamente corretti ad ogni passo.
```

Giusto per non trovarci il giusto commento "Ma non è detto che sia quella giusta!"

## Il manifesto Agile, l'Extreme Programming e il Devops

Negli ultimi decenni sono quindi nate diverse metodologie che incarnano appieno questo approccio, e nel tempo sono diventate capisaldi della nostra professione.

### Il Manifesto Agile

Il primo, temporalmente parlando, è stato il Manifesto Agile. Questo manifesto, stilato nel 2001 e firmato da poco meno di una ventina di professionisti e veterani del settore, fu un modo di rispondere agli attriti e alle difficoltà nella professione del programmatore che da decenni piagavano i progetti software.

Gestiti spesso in modalità "waterfall", i progetti partivano con una fase di design e pianificazione che doveva essere onnicomprensiva, seguita da una fase di sviluppo ed una di test, senza nessun tipo di iterazione; questo approccio portava ad una pianificazione meticolosa ed una stesura di requisiti dettagliattissima, quasi al limite della pedanteria, che però poi si schiantava con la dura realtà durante la fase di sviluppo (o peggio ancora dopo, durante i test) perché, nell'atto dello scrivere codice, quasi sempre emergevano dettagli imprevisti e casi limite che non erano stati contemplati.

> Questo approccio è stato sconvolto e anzi capovolto dal Manifesto Agile, perché porta al centro "gli individui e le interazioni" al posto delle specifiche e delle pianificazioni, incentrando tutto il lavoro sull'ottenere "software funzionante" ad ogni step di sviluppo e non sono alla fine, in modo che sia sempre possibile collaborare assieme e "rispondere al cambiamento", che deve essere accolto sempre, in quanto inevitabile.

Reviewer: ```suggestion
Questo approccio è stato sconvolto e anzi capovolto dal Manifesto Agile, perché porta al centro "gli individui e le interazioni" al posto delle specifiche e delle pianificazioni, incentrando tutto il lavoro sull'ottenere "software funzionante" ad ogni step di sviluppo e non solo alla fine, di modo che sia sempre possibile collaborare assieme e "rispondere al cambiamento", che deve essere accolto sempre, in quanto inevitabile.
```

> Il Manifesto può sembrare, nella sua prima pagina, un po' astratto, ma leggendone meglio il testo si vede come indica tanti modi in cui si può applicare uno strettissimo ciclo del feedback in tutte le fasi della realizzazione di un software: "i cambiamenti ai requisiti vanno accolti", invece che rifiutati perché arrivati tardi, e c'è una forte importanza nel "consegnare software funzionante frequentemente", in modo da dare un riscontro rapido ai committenti; inoltre, "il team deve riflettere ad intervalli regolari" su come il loro lavoro stia procedendo, permettendo al cambiamento di permeare anche le metodologie e le interazioni dei programmatori stessi.

Reviewer: ```suggestion
Il Manifesto può sembrare, nella sua prima pagina, un po' astratto, ma andando più a fondo, questo indica tanti modi in cui si può applicare uno strettissimo ciclo del feedback in tutte le fasi della realizzazione di un software: "i cambiamenti ai requisiti vanno accolti", invece che rifiutati perché arrivati tardi, e viene dato molto rilievo al "consegnare software funzionante frequentemente", in modo da dare e ricevere un riscontro rapido nei confronti dei committenti; inoltre, "il team deve riflettere ad intervalli regolari" su come il loro lavoro stia procedendo, permettendo al cambiamento di permeare anche le metodologie e le interazioni dei programmatori stessi.
```

## L'Extreme Programming e il TDD

> L'Extreme Programming è una delle incarnazioni pratiche del Manifesto Agile. Questa metodologia è stata ideata da Kent Beck nella seconda metà degli anni 90, e anche essa si concentra sul ciclo del feedback, in particolare nel rilascio frequente del software e in una serie di metodologie che permettono al team di sviluppo di iterare e migliorare costantemente il proprio lavoro.

Reviewer: ```suggestion
L'Extreme Programming è una delle incarnazioni pratiche del Manifesto Agile. Questa metodologia è stata ideata da Kent Beck nella seconda metà degli anni '90, e anch'essa si concentra sul ciclo del feedback, in particolare nel rilascio frequente del software e in una serie di metodologie che permettono al team di sviluppo di iterare e migliorare costantemente il proprio lavoro.
```

> Lo strumento principe dell'Extreme Programming è il TDD (Test Driven Development), una metodologia di sviluppo che incarna appieno il concetto di "ciclo del feedback". Può essere riassunta dal fatto che richiede la scrittura di un test prima ancora del codice, e dal ciclo "red-green-refactor":

Reviewer: ```suggestion
Lo strumento principe dell'Extreme Programming è il TDD (Test Driven Development), una metodologia di sviluppo che incarna appieno il concetto di "ciclo del feedback". Può essere riassunta dal fatto che richiede la scrittura di un test prima ancora del codice, e dal ciclo "Red-Green-Refactor":
```

> - **red**: scrivo prima un test "rosso", che descrive prima caratteristica che io desidero dal mio programma, e ne dimostra l'assenza fallendo;

Reviewer: ```suggestion
- **Red**: scrivo prima un test "rosso", che descrive prima caratteristica che io desidero dal mio programma, e ne dimostra l'assenza fallendo;
```

> - **green**: scrivo il pezzo minimo di software per far sì che il test passi, diventando verde;

Reviewer: ```suggestion
- **Green**: scrivo il pezzo minimo di software per far sì che il test passi, diventando verde;
```

> - **refactor**: continuo a rifattorizzare il mio codice, migliorandolo, ma mantenendolo corretto grazie ai test che continuano a rimanere verdi.

Reviewer: ```suggestion
- **Refactor**: Nella fase di refactoring ci troviamo ancora "nel verde", quindi con dei test che passano. In questa fase andremo a migliorare e rifattorizzare il nostro codice per renderlo più performante, manutenibile o scalabile, in base alle necessità e agli obiettivi.
```

> Questo rapido ciclo permette allo sviluppatore di scrivere del codice testato ma soprattutto corretto, che va a sempre a rispettare quanto desiderato, e che facilmente prende la forma necessaria a soddisfare i requisiti descritti nei test. E' opinione di molti poi che il codice scritto in questa modalità sia spesso più ordinato e organizzato, perché la scrittura dei test forza l'architettura del software ad essere più chiara e disaccoppiata.

Reviewer: ```suggestion
Questo rapido ciclo permette allo sviluppatore di scrivere del codice testato ma soprattutto corretto, che va a sempre a rispettare quanto desiderato, e che facilmente prende la forma necessaria a soddisfare i requisiti descritti nei test. È opinione di molti che il codice scritto in questa modalità sia spesso più ordinato e organizzato, perché la scrittura dei test forza l'architettura del software ad essere più chiara e disaccoppiata.

Una nota importante va fatta riguardo a questa metodologia: Pur permettendo di scrivere codice che necessariamente passerà tutti i test scritti in prima istanza, c'è da considerare che non sempre i test scritti rispecchiano le necessità reali, ed è giusto che chi sviluppi tenga in considerazione questa possibilità e non si affidi al TDD come "soluzione ad ogni male".
```

Ho pensato di aggiungere una postilla riguardo al 'rispettare quanto desiderato' nel TDD, che ne pensi?

---
layout: default
title: Ciclo del Feedback
nav_order: 999
---

# Il Ciclo del Feedback

## La flessibilità del software

> Intuitivamente parlando, sembra ovvio che noi sviluppatori, con una connessione internet e un portatile alla mano, possiamo essere enormemente più produttivi di decine di operai da catena di montaggio del secolo scorso. Ma perché è davvero così?

Author: La mia intenzione era porre l'accento su "del secolo scorso", proprio ad indicare che a parità di persona, la tecnologia odierna ha un fattore moltiplicativo enorme sulla produttività del singolo.

Pensi si possa rendere più chiaro questo concetto?

> Intuitivamente parlando, sembra ovvio che noi sviluppatori, con una connessione internet e un portatile alla mano, possiamo essere enormemente più produttivi di decine di operai da catena di montaggio del secolo scorso. Ma perché è davvero così?

Reviewer 2: Non saprei sinceramente, direi semplicemente che il lavoro di uno sviluppatore con il solo uso del computer e della propria conoscenza può avere grandi ripercussioni su un determinato business (sia in termini positivi che negativi eheheh).

> Intuitivamente parlando, sembra ovvio che noi sviluppatori, con una connessione internet e un portatile alla mano, possiamo essere enormemente più produttivi di decine di operai da catena di montaggio del secolo scorso. Ma perché è davvero così?

Reviewer: Mi piace @corradopetrelli , punterei su una considerazione senza confronto, seppur ne comprenda la forza..

> Intuitivamente parlando, sembra ovvio che noi sviluppatori, con una connessione internet e un portatile alla mano, possiamo essere enormemente più produttivi di decine di operai da catena di montaggio del secolo scorso. Ma perché è davvero così?

Reviewer: Ciao @Jean85 , quando sei riuscito a rivedere quest'ultimo pezzo ripartiamo con le review!

> Intuitivamente parlando, sembra ovvio che noi sviluppatori, con una connessione internet e un portatile alla mano, possiamo essere enormemente più produttivi di decine di operai da catena di montaggio del secolo scorso. Ma perché è davvero così?

Reviewer: ```suggestion
Intuitivamente parlando, sembra ovvio che noi sviluppatori, con una connessione internet e un portatile alla mano, siamo in grado di portare a termine qualsiasi tipo di attività in tempi rapidi e con un impatto enorme sul mondo che ci circonda. Ma come è possibile?
```

@Jean85 @corradopetrelli ho provato a riformulare!

> Sicuramente un fattore è la possibilità di lavorare con strumenti frutto di decenni di evoluzione tecnologica esponenziale, ma la vera ragione, il vero discriminante, sta nel prefisso "soft" nella parola "software": il prodotto del nostro lavoro è spesso intangibile e facilmente manipolabile, tanto che se commettiamo un errore, non dobbiamo prendere il semilavorato e buttarlo via, sprecando materie prime, ma nella peggiore delle ipotesi abbiamo solo sprecato un po' di tempo.

Reviewer 3: manipolabile, tanto che se commettiamo un errore, -> manipolabile tanto che, se commettiamo un errore,

sprecando materie prime, ma nella peggiore delle ipotesi abbiamo solo sprecato un po' di tempo  -> sprecando materie prime ma, nella peggiore delle ipotesi, abbiamo solo sprecato un po' di tempo

Author: Fatto in fix: incorpora suggerimenti di review Alessandro Lai Moments ago

- **Refactor**: Nella fase di refactoring ci troviamo ancora "nel verde", quindi con dei test che passano. In questa fase andremo a migliorare e rifattorizzare il nostro codice per renderlo più performante, manutenibile o scalabile, in base alle necessità e agli obiettivi.

Questo rapido ciclo permette allo sviluppatore di scrivere del codice testato ma soprattutto corretto, che va a sempre a rispettare quanto desiderato, e che facilmente prende la forma necessaria a soddisfare i requisiti descritti nei test. È opinione di molti che il codice scritto in questa modalità sia spesso più ordinato e organizzato, perché la scrittura dei test forza l'architettura del software ad essere più chiara e disaccoppiata.

Una nota importante va fatta riguardo a questa metodologia: Pur permettendo di scrivere codice che necessariamente passerà tutti i test scritti in prima istanza, c'è da considerare che non sempre i test scritti rispecchiano le necessità reali, ed è giusto che chi sviluppi tenga in considerazione questa possibilità e non si affidi al TDD come "soluzione ad ogni male".

## Conclusione

> Come abbiamo visto, alle fondamenta di tutti questi validi approcci si trova **il ciclo del feedback**: vista la malleabilità del software e la relativa semplicità con cui si può evolvere e sperimentare, avere un ciclo del feedback breve ed efficace si è rapidamente dimostrato lo strumento vincente per uscire da quell'impasse. Nessuno di queste metodologie è però un dettame rigido e scolpito nella pietra: si tratta più di pensieri ed esperienze collettive che emergono organicamente dalla comunità degli sviluppatori. Spesso tra di essi vi è commistione e manca un confine rigido: tanti team al giorno d'oggi infatti tendono a far tesoro di tutte queste metodologie e ad implementarle in maniera "personalizzata", andando a cucirsi su misura un approccio che incarna appieno il pensiero del Manifesto Agile: "Le architetture, i requisiti e la progettazione migliori emergono da team che si auto-organizzano.".

Author: Ho preferito riformulare il finale della frase dandogli una connotazione più positiva; ti convince?

> Come abbiamo visto, alle fondamenta di tutti questi validi approcci si trova **il ciclo del feedback**: vista la malleabilità del software e la relativa semplicità con cui si può evolvere e sperimentare, avere un ciclo del feedback breve ed efficace si è rapidamente dimostrato lo strumento vincente per uscire da quell'impasse. Nessuno di queste metodologie è però un dettame rigido e scolpito nella pietra: si tratta più di pensieri ed esperienze collettive che emergono organicamente dalla comunità degli sviluppatori. Spesso tra di essi vi è commistione e manca un confine rigido: tanti team al giorno d'oggi infatti tendono a far tesoro di tutte queste metodologie e ad implementarle in maniera "personalizzata", andando a cucirsi su misura un approccio che incarna appieno il pensiero del Manifesto Agile: "Le architetture, i requisiti e la progettazione migliori emergono da team che si auto-organizzano.".

Reviewer 2: Sì 😄

Tutto questo non è nient'altro che la "Terza rivoluzione industriale", come viene chiamata nei libri di scuola; dopo alcuni decenni in cui il collo di bottiglia è stata la potenza di calcolo, al giorno d'oggi abbiamo a disposizione CPU e memoria sufficienti per qualunque task mondano, ma questa pura forza bruta ha esposto una nuova tipologia di problemi, quelli organizzativi: far lavorare decine, se non centinaia di sviluppatori assieme sullo stesso progetto software è un problema non banale, e che tanti prima di noi si sono trovati a dover risolvere.

Uno strumento che è emerso vincitore da queste difficoltà è il **ciclo del feedback**; ma di che si tratta? È un approccio in cui, a seguito di un'azione, si va a misurarne rapidamente l'effetto (il "feedback"), così da ottenerne nuove informazioni, che saranno impiegate per intraprendere i prossimi passi in maniera più conscia e informata, creando così un ciclo continuo. Questo generalmente permette di procedere in un'attività a piccoli passi ma con una discreta rapidità e, cosa ancora più importante, solitamente nella giusta direzione, visto che eventuali errori di rotta possono essere rapidamente corretti ad ogni passo.

> ## Il manifesto Agile, l'Extreme Programming e il Devops

Reviewer: @Jean85 mi sono appena reso conto che qui citiamo DevOps ma poi non ne parliamo. Che ne pensi di un ultimo slot di spiegazione?

Io proporrei una cosa del tipo: Dalla nascita della metodologia Agile, molte altre filosofie di sviluppo hanno preso alcuni punti del manifesto e li hanno raffinati e sviluppati nel proprio ramo di competenza. Una di queste filosofie è quella DevOps, che ha come capi saldi... I vantaggi principali sono...

Che ne dici?

Author: @Cadienvan aggiunto in 03a36c8, vai di review! :smile:

Lo strumento principe dell'Extreme Programming è il TDD (Test Driven Development), una metodologia di sviluppo che incarna appieno il concetto di "ciclo del feedback". Può essere riassunta dal fatto che richiede la scrittura di un test prima ancora del codice, e dal ciclo "Red-Green-Refactor":

- **Red**: Lo scopo di questa fase è scrivere un test che informi circa l'implementazione di una funzionalità. La prova verrà superata solo quando le sue aspettative saranno soddisfatte. In mancanza di codice a supporto, naturalmente, il test fallirà.
- **Green**: In questa fase, lo scopo è implementare il codice necessario per superare il test. L'obiettivo è trovare una soluzione, senza preoccuparsi di ottimizzarne l'implementazione.
- **Refactor**: Nella fase di refactoring ci troviamo ancora "nel verde", quindi con dei test che passano. In questa fase andremo a migliorare e rifattorizzare il nostro codice per renderlo più performante, manutenibile o scalabile, in base alle necessità e agli obiettivi.

> Questo rapido ciclo permette allo sviluppatore di scrivere del codice testato ma soprattutto corretto, che va a sempre a rispettare quanto desiderato, e che facilmente prende la forma necessaria a soddisfare i requisiti descritti nei test. È opinione di molti che il codice scritto in questa modalità sia spesso più ordinato e organizzato, perché la scrittura dei test forza l'architettura del software ad essere più chiara e disaccoppiata.

Reviewer 4: Typo

```suggestion
Questo rapido ciclo permette allo sviluppatore di scrivere del codice testato ma soprattutto corretto, che va sempre a rispettare quanto desiderato, e che facilmente prende la forma necessaria a soddisfare i requisiti descritti nei test. È opinione di molti che il codice scritto in questa modalità sia spesso più ordinato e organizzato, perché la scrittura dei test forza l'architettura del software ad essere più chiara e disaccoppiata.
```

---
layout: default

> title: Il Ciclo del Feedback nello sviluppo software

Reviewer: ```suggestion
title: Metodologie Agile e ciclo del feedback
```

- **Refactor**: Nella fase di refactoring ci troviamo ancora "nel verde", quindi con dei test che passano. In questa fase andremo a migliorare e rifattorizzare il nostro codice per renderlo più performante, manutenibile o scalabile, in base alle necessità e agli obiettivi.

Questo rapido ciclo permette allo sviluppatore di scrivere del codice testato ma soprattutto corretto, che va sempre a rispettare quanto desiderato, e che facilmente prende la forma necessaria a soddisfare i requisiti descritti nei test. È opinione di molti che il codice scritto in questa modalità sia spesso più ordinato e organizzato, perché la scrittura dei test forza l'architettura del software ad essere più chiara e disaccoppiata.

È fondamentale sottolineare un aspetto cruciale riguardo a questa metodologia: sebbene consenta la scrittura di codice in grado di superare tutti i test inizialmente definiti, è importante considerare che questi potrebbero non sempre rispecchiare appieno le esigenze reali. Pertanto, chi sviluppa dovrebbe essere consapevole di questa possibilità e non fare dell'approccio Test-Driven Development la "soluzione ad ogni male".

> ## L'Extreme Programming e il TDD

Reviewer: ```suggestion
## DevOps
```

Dalla nascita della metodologia Agile, molte altre filosofie di sviluppo hanno preso alcuni punti del manifesto e li hanno raffinati e sviluppati nel proprio ramo di competenza. Una di queste filosofie è quella DevOps, nata per risolvere l'annoso problema della separazione tra sviluppatori (dev) e sistemisti (ops): in passato spesso questi due ruoli rimanevano fortemente separati, con gli sviluppatori che "scaricavano" le modifiche software ai sistemisti, che si dovevano poi preoccupare di portarle in produzione, senza conoscerle appieno ma al tempo stesso avendo la responsabilità della loro stabilità e resilienza. Questa separazione è l'esempio perfetto di un ciclo del feedback lento e piagato da troppi passaggi di consegne.

> Questa metodologia nasce come risposta a questa netta separazione, creando un ruolo "ibrido" in cui si ha la mescolanza delle due competenze; ha come capi saldi la condivisione delle responsabilità (tra dev e ops), l'automazione e (per l'appunto) un rapido ciclo di feedback. Nella pratica, applicare i principi del DevOps spesso si sintetizza nell'automatizzare il flusso di rilascio in produzione e nel mantenere la resposabilità del deploy all'interno del team che sviluppa il software stesso, così che gli sviluppatori possano beneficiare istantaneamente del feedback generato dal rilascio del loro software in ambiente di produzione reale, e possano intervenire tempestivamente (e con la giusta dose di conoscenze) in caso di eventuali difetti.

Reviewer: ```suggestion
Questa metodologia nasce come risposta a questa netta separazione, creando un ruolo "ibrido" in cui si ha la mescolanza delle due competenze; ha come capi saldi la condivisione delle responsabilità (tra dev e ops), l'automazione e (per l'appunto) un rapido ciclo di feedback.

Nella pratica, applicare i principi del DevOps spesso si sintetizza nell'automatizzare il flusso di rilascio in produzione e nel mantenere la responsabilità del deploy all'interno del team che sviluppa il software stesso, così che chi sviluppa possa beneficiare istantaneamente del feedback generato dal rilascio del loro software in ambiente di produzione reale, e possano intervenire tempestivamente (e con la giusta dose di conoscenze) in caso di eventuali difetti.
```

---
layout: default
title: Metodologie Agile e ciclo del feedback
nav_order: 10
---

<!-- prettier-ignore-start -->
# Il Ciclo del Feedback
{: .no_toc }

- TOC
{:toc}

<!-- prettier-ignore-end -->

## La flessibilità del software

Intuitivamente parlando, sembra ovvio che noi sviluppatori, con una connessione internet e un portatile alla mano, siamo in grado di portare a termine qualsiasi tipo di attività in tempi rapidi e con un impatto enorme sul mondo che ci circonda. Ma come è possibile?

Sicuramente un fattore è la possibilità di lavorare con strumenti frutto di decenni di evoluzione tecnologica esponenziale ma, la vera ragione, il vero discriminante, sta nel prefisso "soft" nella parola "software": il prodotto del nostro lavoro è spesso intangibile e facilmente manipolabile tanto che, se commettiamo un errore, non dobbiamo prendere il semilavorato e buttarlo via, sprecando materie prime ma, nella peggiore delle ipotesi, abbiamo solo sprecato un po' di tempo.

> Tutto questo non è nient'altro che la "Terza rivoluzione industriale", come viene chiamata nei libri di scuola; dopo alcuni decenni in cui il collo di bottiglia è stata la potenza di calcolo, al giorno d'oggi abbiamo a disposizione CPU e memoria sufficienti per qualunque task mondano; ma questa pura forza bruta ha esposto una nuova tipologia di problemi, quelli organizzativi: far lavorare decine, se non centinaia, di sviluppatori assieme sullo stesso progetto software è un problema non banale, che tanti prima di noi si sono trovati a dover risolvere.

Reviewer 5: In che senso task mondano?

Author: Credo sia un modo rapido per far capire che non parliamo di cose astruse tipo simulazioni di fisica quantistica :smile:

Uno strumento che è emerso vincitore da queste difficoltà è il **ciclo del feedback**; ma di che si tratta? È un approccio in cui, a seguito di un'azione, si va a misurarne rapidamente l'effetto (il "feedback"), così da ottenerne nuove informazioni, che saranno impiegate per intraprendere i prossimi passi in maniera più conscia e informata, creando così un ciclo continuo. Questo generalmente permette di procedere in un'attività a piccoli passi ma con una discreta rapidità e, cosa ancora più importante, solitamente nella giusta direzione, visto che eventuali errori di rotta possono essere rapidamente corretti ad ogni passo.

## Il manifesto Agile, l'Extreme Programming e il Devops

Negli ultimi decenni sono quindi nate diverse metodologie che incarnano appieno questo approccio, e nel tempo sono diventate capisaldi della nostra professione.

### Il Manifesto Agile

Il primo, temporalmente parlando, è stato il Manifesto Agile. Questo manifesto, stilato nel 2001 e firmato da poco meno di una ventina di professionisti e veterani del settore, fu un modo di rispondere agli attriti e alle difficoltà nella professione del programmatore che da decenni piagavano i progetti software.

> Gestiti spesso in modalità "waterfall", i progetti partivano con una fase di design e pianificazione che doveva essere onnicomprensiva, seguita da una fase di sviluppo ed una di test, senza nessun tipo di iterazione; questo approccio portava ad una pianificazione meticolosa ed una stesura di requisiti dettagliattissima, quasi al limite della pedanteria, che però poi si schiantava con la dura realtà durante la fase di sviluppo (o peggio ancora dopo, durante i test) perché, nell'atto dello scrivere codice, quasi sempre emergevano dettagli imprevisti e casi limite che non erano stati contemplati.

Reviewer 3: ```suggestion
Gestiti spesso in modalità "waterfall", i progetti partivano con una fase di design e pianificazione che doveva essere onnicomprensiva, seguita da una fase di sviluppo ed una di test, senza nessun tipo di iterazione; questo approccio portava ad una pianificazione meticolosa ed una stesura di requisiti dettagliatissima, quasi al limite della pedanteria, che però poi si schiantava con la dura realtà durante la fase di sviluppo (o peggio ancora dopo, durante i test) perché, nell'atto dello scrivere codice, quasi sempre emergevano dettagli imprevisti e casi limite che non erano stati contemplati.
```

Questo approccio è stato sconvolto e anzi capovolto dal Manifesto Agile, perché porta al centro "gli individui e le interazioni" al posto delle specifiche e delle pianificazioni, incentrando tutto il lavoro sull'ottenere "software funzionante" ad ogni step di sviluppo e non solo alla fine, di modo che sia sempre possibile collaborare assieme e "rispondere al cambiamento", che deve essere accolto sempre, in quanto inevitabile.

> Il Manifesto può sembrare, nella sua prima pagina, un po' astratto, ma andando più a fondo, questo indica tanti modi in cui si può applicare uno strettissimo ciclo del feedback in tutte le fasi della realizzazione di un software: "i cambiamenti ai requisiti vanno accolti", invece che rifiutati perché arrivati tardi, e viene dato molto rilievo al "consegnare software funzionante frequentemente", in modo da dare e ricevere un riscontro rapido nei confronti dei committenti; inoltre, "il team deve riflettere ad intervalli regolari" su come il loro lavoro stia procedendo, permettendo al cambiamento di permeare anche le metodologie e le interazioni dei programmatori stessi.

Reviewer 3: ```suggestion
Il Manifesto può sembrare, nella sua prima pagina, un po' astratto ma, andando più a fondo, questo indica tanti modi in cui si può applicare uno strettissimo ciclo del feedback in tutte le fasi della realizzazione di un software: "i cambiamenti ai requisiti vanno accolti", invece che rifiutati perché arrivati tardi, e viene dato molto rilievo al "consegnare software funzionante frequentemente", in modo da dare e ricevere un riscontro rapido nei confronti dei committenti; inoltre, "il team deve riflettere ad intervalli regolari" su come il loro lavoro stia procedendo, permettendo al cambiamento di permeare anche le metodologie e le interazioni dei programmatori stessi.
```

## L'Extreme Programming e il TDD

L'Extreme Programming è una delle incarnazioni pratiche del Manifesto Agile. Questa metodologia è stata ideata da Kent Beck nella seconda metà degli anni '90, e anch'essa si concentra sul ciclo del feedback, in particolare nel rilascio frequente del software e in una serie di metodologie che permettono al team di sviluppo di iterare e migliorare costantemente il proprio lavoro.

> Lo strumento principe dell'Extreme Programming è il TDD (Test Driven Development), una metodologia di sviluppo che incarna appieno il concetto di "ciclo del feedback". Può essere riassunta dal fatto che richiede la scrittura di un test prima ancora del codice, e dal ciclo "Red-Green-Refactor":

Reviewer 5: Incarna è già stato usato nel paragrafo precedente e non mi suona benissimo. Se non si trova un sinonimo fa nulla ma penso possa fare la differenza.

Author: "concretizza"? "realizza"?

Reviewer: "concretizza il ciclo del feedback" mi piace molto!

Author: Fixed in 91dd339

- **Red**: Lo scopo di questa fase è scrivere un test che informi circa l'implementazione di una funzionalità. La prova verrà superata solo quando le sue aspettative saranno soddisfatte. In mancanza di codice a supporto, naturalmente, il test fallirà.
- **Green**: In questa fase, lo scopo è implementare il codice necessario per superare il test. L'obiettivo è trovare una soluzione, senza preoccuparsi di ottimizzarne l'implementazione.
- **Refactor**: Nella fase di refactoring ci troviamo ancora "nel verde", quindi con dei test che passano. In questa fase andremo a migliorare e rifattorizzare il nostro codice per renderlo più performante, manutenibile o scalabile, in base alle necessità e agli obiettivi.

> Questo rapido ciclo permette allo sviluppatore di scrivere del codice testato ma soprattutto corretto, che va sempre a rispettare quanto desiderato, e che facilmente prende la forma necessaria a soddisfare i requisiti descritti nei test. È opinione di molti che il codice scritto in questa modalità sia spesso più ordinato e organizzato, perché la scrittura dei test forza l'architettura del software ad essere più chiara e disaccoppiata.

Reviewer 5: ```suggestion
Questo rapido ciclo permette allo sviluppatore di scrivere del codice testato ma soprattutto corretto, che va sempre a rispettare quanto desiderato, e che facilmente prende la forma necessaria a soddisfare i requisiti descritti nei test. È opinione di molti che il codice scritto in questa modalità sia spesso più ordinato ed organizzato, perché la scrittura dei test forza l'architettura del software ad essere più chiara e disaccoppiata.
```

Questa metodologia nasce come risposta a questa netta separazione, creando un ruolo "ibrido" in cui si ha la mescolanza delle due competenze; ha come capi saldi la condivisione delle responsabilità (tra dev e ops), l'automazione e (per l'appunto) un rapido ciclo di feedback.

Nella pratica, applicare i principi del DevOps spesso si sintetizza nell'automatizzare il flusso di rilascio in produzione e nel mantenere la responsabilità del deploy all'interno del team che sviluppa il software stesso, così che chi sviluppa possa beneficiare istantaneamente del feedback generato dal rilascio del loro software in ambiente di produzione reale, e possano intervenire tempestivamente (e con la giusta dose di conoscenze) in caso di eventuali difetti.

## Conclusione

> Come abbiamo visto, alle fondamenta di tutti questi validi approcci si trova **il ciclo del feedback**: vista la malleabilità del software e la relativa semplicità con cui si può evolvere e sperimentare, avere un ciclo del feedback breve ed efficace si è rapidamente dimostrato lo strumento vincente per avanzare rapidamente e produrre subito valore per l'utente finale. Nessuno di queste metodologie è però un dettame rigido e scolpito nella pietra: si tratta più di pensieri ed esperienze collettive che emergono organicamente dalla comunità degli sviluppatori. Spesso tra di essi vi è commistione e manca un confine rigido: tanti team al giorno d'oggi infatti tendono a far tesoro di tutte queste metodologie e ad implementarle in maniera "personalizzata", andando a cucirsi su misura un approccio che incarna appieno il pensiero del Manifesto Agile: "Le architetture, i requisiti e la progettazione migliori emergono da team che si auto-organizzano.".

Reviewer 6: ```suggestion
Come abbiamo visto, alle fondamenta di tutti questi validi approcci si trova **il ciclo del feedback**: vista la malleabilità del software e la relativa semplicità con cui si può evolvere e sperimentare, avere un ciclo del feedback breve ed efficace si è dimostrato lo strumento vincente per avanzare rapidamente e produrre subito valore per l'utente finale. Nessuna di queste metodologie è però un dettame rigido e scolpito nella pietra: si tratta più di pensieri ed esperienze collettive che emergono organicamente dalla comunità degli sviluppatori. Spesso tra di essi vi è commistione e manca un confine rigido: tanti team al giorno d'oggi infatti tendono a far tesoro di tutte queste metodologie e ad implementarle in maniera "personalizzata", andando a cucirsi su misura un approccio che incarna appieno il pensiero del Manifesto Agile: "Le architetture, i requisiti e la progettazione migliori emergono da team che si auto-organizzano.".
```
